# Исследование надежности заемщиков


Во второй части проекта вы выполните шаги 3 и 4. Их вручную проверит ревьюер.
Чтобы вам не пришлось писать код заново для шагов 1 и 2, мы добавили авторские решения в ячейки с кодом. 



## Откройте таблицу и изучите общую информацию о данных

**Задание 1. Импортируйте библиотеку pandas. Считайте данные из csv-файла в датафрейм и сохраните в переменную `data`. Путь к файлу:**

`/datasets/data.csv`

In [1]:
import pandas as pd

try:
    data = pd.read_csv('/datasets/data.csv')
except:
    data = pd.read_csv('https://code.s3.yandex.net/datasets/data.csv')

**Задание 2. Выведите первые 20 строчек датафрейма `data` на экран.**

In [2]:
data.head(20)

children  days_employed  dob_years            education  education_id  \
0          1   -8437.673028         42               высшее             0   
1          1   -4024.803754         36              среднее             1   
2          0   -5623.422610         33              Среднее             1   
3          3   -4124.747207         32              среднее             1   
4          0  340266.072047         53              среднее             1   
5          0    -926.185831         27               высшее             0   
6          0   -2879.202052         43               высшее             0   
7          0    -152.779569         50              СРЕДНЕЕ             1   
8          2   -6929.865299         35               ВЫСШЕЕ             0   
9          0   -2188.756445         41              среднее             1   
10         2   -4171.483647         36               высшее             0   
11         0    -792.701887         40              среднее             1   
12         0            NaN         65              среднее             1   
13         0   -1846.641941         54  неоконченное высшее             2   
14         0   -1844.956182         56               высшее             0   
15         1    -972.364419         26              среднее             1   
16         0   -1719.934226         35              среднее             1   
17         0   -2369.999720         33               высшее             0   
18         0  400281.136913         53              среднее             1   
19         0  -10038.818549         48              СРЕДНЕЕ             1   

       family_status  family_status_id gender income_type  debt  \
0    женат / замужем                 0      F   сотрудник     0   
1    женат / замужем                 0      F   сотрудник     0   
2    женат / замужем                 0      M   сотрудник     0   
3    женат / замужем                 0      M   сотрудник     0   
4   гражданский брак                 1      F   пенсионер     0   
5   гражданский брак                 1      M   компаньон     0   
6    женат / замужем                 0      F   компаньон     0   
7    женат / замужем                 0      M   сотрудник     0   
8   гражданский брак                 1      F   сотрудник     0   
9    женат / замужем                 0      M   сотрудник     0   
10   женат / замужем                 0      M   компаньон     0   
11   женат / замужем                 0      F   сотрудник     0   
12  гражданский брак                 1      M   пенсионер     0   
13   женат / замужем                 0      F   сотрудник     0   
14  гражданский брак                 1      F   компаньон     1   
15   женат / замужем                 0      F   сотрудник     0   
16   женат / замужем                 0      F   сотрудник     0   
17  гражданский брак                 1      M   сотрудник     0   
18    вдовец / вдова                 2      F   пенсионер     0   
19         в разводе                 3      F   сотрудник     0   

     total_income                                 purpose  
0   253875.639453                           покупка жилья  
1   112080.014102                 приобретение автомобиля  
2   145885.952297                           покупка жилья  
3   267628.550329              дополнительное образование  
4   158616.077870                         сыграть свадьбу  
5   255763.565419                           покупка жилья  
6   240525.971920                       операции с жильем  
7   135823.934197                             образование  
8    95856.832424                   на проведение свадьбы  
9   144425.938277                 покупка жилья для семьи  
10  113943.491460                    покупка недвижимости  
11   77069.234271       покупка коммерческой недвижимости  
12            NaN                         сыграть свадьбу  
13  130458.228857                 приобретение автомобиля  
14  165127.911772              покупка жилой недвижимости  
15  116820.904450 

**Задание 3. Выведите основную информацию о датафрейме с помощью метода `info()`.**

In [3]:
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 21525 entries, 0 to 21524
Data columns (total 12 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   children          21525 non-null  int64  
 1   days_employed     19351 non-null  float64
 2   dob_years         21525 non-null  int64  
 3   education         21525 non-null  object 
 4   education_id      21525 non-null  int64  
 5   family_status     21525 non-null  object 
 6   family_status_id  21525 non-null  int64  
 7   gender            21525 non-null  object 
 8   income_type       21525 non-null  object 
 9   debt              21525 non-null  int64  
 10  total_income      19351 non-null  float64
 11  purpose           21525 non-null  object 
dtypes: float64(2), int64(5), object(5)
memory usage: 2.0+ MB


## Предобработка данных

### Удаление пропусков

**Задание 4. Выведите количество пропущенных значений для каждого столбца. Используйте комбинацию двух методов.**

In [4]:
data.isna().sum()

children               0
days_employed       2174
dob_years              0
education              0
education_id           0
family_status          0
family_status_id       0
gender                 0
income_type            0
debt                   0
total_income        2174
purpose                0
dtype: int64

**Задание 5. В двух столбцах есть пропущенные значения. Один из них — `days_employed`. Пропуски в этом столбце вы обработаете на следующем этапе. Другой столбец с пропущенными значениями — `total_income` — хранит данные о доходах. На сумму дохода сильнее всего влияет тип занятости, поэтому заполнить пропуски в этом столбце нужно медианным значением по каждому типу из столбца `income_type`. Например, у человека с типом занятости `сотрудник` пропуск в столбце `total_income` должен быть заполнен медианным доходом среди всех записей с тем же типом.**

In [5]:
for t in data['income_type'].unique():
    data.loc[(data['income_type'] == t) & (data['total_income'].isna()), 'total_income'] = \
    data.loc[(data['income_type'] == t), 'total_income'].median()
    

### Обработка аномальных значений

**Задание 6. В данных могут встречаться артефакты (аномалии) — значения, которые не отражают действительность и появились по какой-то ошибке. таким артефактом будет отрицательное количество дней трудового стажа в столбце `days_employed`. Для реальных данных это нормально. Обработайте значения в этом столбце: замените все отрицательные значения положительными с помощью метода `abs()`.**

In [6]:
data['days_employed'] = data['days_employed'].abs()

**Задание 7. Для каждого типа занятости выведите медианное значение трудового стажа `days_employed` в днях.**

In [7]:
data.groupby('income_type')['days_employed'].agg('median')

income_type
безработный        366413.652744
в декрете            3296.759962
госслужащий          2689.368353
компаньон            1547.382223
пенсионер          365213.306266
предприниматель       520.848083
сотрудник            1574.202821
студент               578.751554
Name: days_employed, dtype: float64

У двух типов (безработные и пенсионеры) получатся аномально большие значения. Исправить такие значения сложно, поэтому оставьте их как есть. Тем более этот столбец не понадобится вам для исследования.

**Задание 8. Выведите перечень уникальных значений столбца `children`.**

In [8]:
data['children'].unique()

array([ 1,  0,  3,  2, -1,  4, 20,  5])

**Задание 9. В столбце `children` есть два аномальных значения. Удалите строки, в которых встречаются такие аномальные значения из датафрейма `data`.**

In [9]:
data = data[(data['children'] != -1) & (data['children'] != 20)]

**Задание 10. Ещё раз выведите перечень уникальных значений столбца `children`, чтобы убедиться, что артефакты удалены.**

In [10]:
data['children'].unique()

array([1, 0, 3, 2, 4, 5])

### Удаление пропусков (продолжение)

**Задание 11. Заполните пропуски в столбце `days_employed` медианными значениями по каждого типа занятости `income_type`.**

In [11]:
for t in data['income_type'].unique():
    data.loc[(data['income_type'] == t) & (data['days_employed'].isna()), 'days_employed'] = \
    data.loc[(data['income_type'] == t), 'days_employed'].median()

**Задание 12. Убедитесь, что все пропуски заполнены. Проверьте себя и ещё раз выведите количество пропущенных значений для каждого столбца с помощью двух методов.**

In [12]:
data.isna().sum()

children            0
days_employed       0
dob_years           0
education           0
education_id        0
family_status       0
family_status_id    0
gender              0
income_type         0
debt                0
total_income        0
purpose             0
dtype: int64

### Изменение типов данных

**Задание 13. Замените вещественный тип данных в столбце `total_income` на целочисленный с помощью метода `astype()`.**

In [13]:
data['total_income'] = data['total_income'].astype(int)

### Обработка дубликатов

**Задание 14. Обработайте неявные дубликаты в столбце `education`. В этом столбце есть одни и те же значения, но записанные по-разному: с использованием заглавных и строчных букв. Приведите их к нижнему регистру. Проверьте остальные столбцы.**

In [14]:
data['education'] = data['education'].str.lower()

**Задание 15. Выведите на экран количество строк-дубликатов в данных. Если такие строки присутствуют, удалите их.**

In [15]:
data.duplicated().sum()

71

In [16]:
data = data.drop_duplicates()

### Категоризация данных

**Задание 16. На основании диапазонов, указанных ниже, создайте в датафрейме `data` столбец `total_income_category` с категориями:**

- 0–30000 — `'E'`;
- 30001–50000 — `'D'`;
- 50001–200000 — `'C'`;
- 200001–1000000 — `'B'`;
- 1000001 и выше — `'A'`.


**Например, кредитополучателю с доходом 25000 нужно назначить категорию `'E'`, а клиенту, получающему 235000, — `'B'`. Используйте собственную функцию с именем `categorize_income()` и метод `apply()`.**

In [17]:
def categorize_income(income):
    try:
        if 0 <= income <= 30000:
            return 'E'
        elif 30001 <= income <= 50000:
            return 'D'
        elif 50001 <= income <= 200000:
            return 'C'
        elif 200001 <= income <= 1000000:
            return 'B'
        elif income >= 1000001:
            return 'A'
    except:
        pass

In [18]:
data['total_income_category'] = data['total_income'].apply(categorize_income)

**Задание 17. Выведите на экран перечень уникальных целей взятия кредита из столбца `purpose`.**

In [19]:
data['purpose'].unique()

array(['покупка жилья', 'приобретение автомобиля',
       'дополнительное образование', 'сыграть свадьбу',
       'операции с жильем', 'образование', 'на проведение свадьбы',
       'покупка жилья для семьи', 'покупка недвижимости',
       'покупка коммерческой недвижимости', 'покупка жилой недвижимости',
       'строительство собственной недвижимости', 'недвижимость',
       'строительство недвижимости', 'на покупку подержанного автомобиля',
       'на покупку своего автомобиля',
       'операции с коммерческой недвижимостью',
       'строительство жилой недвижимости', 'жилье',
       'операции со своей недвижимостью', 'автомобили',
       'заняться образованием', 'сделка с подержанным автомобилем',
       'получение образования', 'автомобиль', 'свадьба',
       'получение дополнительного образования', 'покупка своего жилья',
       'операции с недвижимостью', 'получение высшего образования',
       'свой автомобиль', 'сделка с автомобилем',
       'профильное образование', 'высшее об

**Задание 18. Создайте функцию, которая на основании данных из столбца `purpose` сформирует новый столбец `purpose_category`, в который войдут следующие категории:**

- `'операции с автомобилем'`,
- `'операции с недвижимостью'`,
- `'проведение свадьбы'`,
- `'получение образования'`.

**Например, если в столбце `purpose` находится подстрока `'на покупку автомобиля'`, то в столбце `purpose_category` должна появиться строка `'операции с автомобилем'`.**

**Используйте собственную функцию с именем `categorize_purpose()` и метод `apply()`. Изучите данные в столбце `purpose` и определите, какие подстроки помогут вам правильно определить категорию.**

In [20]:
def categorize_purpose(row):
    try:
        if 'автом' in row:
            return 'операции с автомобилем'
        elif 'жил' in row or 'недвиж' in row:
            return 'операции с недвижимостью'
        elif 'свад' in row:
            return 'проведение свадьбы'
        elif 'образов' in row:
            return 'получение образования'
    except:
        return 'нет категории'

In [21]:
data['purpose_category'] = data['purpose'].apply(categorize_purpose)

### Шаг 3. Исследуйте данные и ответьте на вопросы

#### 3.1 Есть ли зависимость между количеством детей и возвратом кредита в срок?

In [22]:
# Ваш код будет здесь. Вы можете создавать новые ячейки.
debt_children = data.groupby("children")["debt"].mean()
data.groupby("children").agg({"debt":["count","mean"]})

debt          
          count      mean
children                 
0         14091  0.075438
1          4808  0.092346
2          2052  0.094542
3           330  0.081818
4            41  0.097561
5             9  0.000000

In [23]:
print(f"Случаи задолжностей к погашениям без задолжностей без детей в семье: {('%.2f' % (debt_children[0] * 100))}%")
print(f"Случаи задолжностей к погашениям без задолжностей c одним ребенком в семье: {('%.2f' % (debt_children[1] * 100))}%")
print(f"Случаи задолжностей к погашениям без задолжностей с двумя детьми в семье: {('%.2f' % (debt_children[2] * 100))}%")
print(f"Случаи задолжностей к погашениям без задолжностей с тремя  детьми в семье: {('%.2f' % (debt_children[3] * 100))}%")
print(f"Случаи задолжностей к погашениям без задолжностей с четырьмя детьми в семье: {('%.2f' % (debt_children[4] * 100))}%")
print(f"Случаи задолжностей к погашениям без задолжностей с пятью детьми в семье: {('%.2f' % (debt_children[5] * 100))}%")

Случаи задолжностей к погашениям без задолжностей без детей в семье: 7.54%
Случаи задолжностей к погашениям без задолжностей c одним ребенком в семье: 9.23%
Случаи задолжностей к погашениям без задолжностей с двумя детьми в семье: 9.45%
Случаи задолжностей к погашениям без задолжностей с тремя  детьми в семье: 8.18%
Случаи задолжностей к погашениям без задолжностей с четырьмя детьми в семье: 9.76%
Случаи задолжностей к погашениям без задолжностей с пятью детьми в семье: 0.00%


**Вывод:** Что могу сказать точно, то для вывода о зависимости кол-ва детей и возвратом кредита в срок, для случаев, когда детей в семье 4 и 5, не хватает данных.
Если же брать в целом, то наличие детей в семье всё же действительно сказывается на проблемы возврата кредита в срок - касаемо случаев, когда в семья 1-2 ребенка, данные отличаются не сильно.
Что касаемо случаев, когда в семье 3 ребенка, то процент имел ли задолженность к погашениям без задолжностей сильно ниже, что вероятнее всего можно списать на недостаток данных, ибо таких случаев всего 330, что недостаточно для качественного анализа.
Еще большая проблема в кол-ве данных для 4 и 5 детей в семье.
Самый же низший процент случаев задолжностей к погашениям без задолжностей - именно без детей вовсе.
Но если отбросить проблемы качественного анализа, то чаще всего кредит выплачивают в семьях с 5 детьми (недостаток данных, всего 9 семей)


#### 3.2 Есть ли зависимость между семейным положением и возвратом кредита в срок?

In [24]:
data_married = pd.pivot_table(data, index=["family_status_id", "family_status"], values=["debt"])
data_married1 = data.groupby("family_status_id")["debt"].mean()
#data_married["debt"][2][0]

In [25]:
print(f"Случаи задолжностей к погашениям без задолжностей для женатых: {('%.2f' % (data_married1[0] * 100))}%")
print(f"Случаи задолжностей к погашениям без задолжностей в гражданском браке: {('%.2f' % (data_married1[1] * 100))}%")
print(f"Случаи задолжностей к погашениям без задолжностей для вдов и вдовцев : {('%.2f' % (data_married1[2] * 100))}%")
print(f"Случаи задолжностей к погашениям без задолжностей для тех, кто в разводе: {('%.2f' % (data_married1[3] * 100))}%")
print(f"Случаи задолжностей к погашениям без задолжностей для не женатых/не замужних: {('%.2f' % (data_married1[4] * 100))}%")


Случаи задолжностей к погашениям без задолжностей для женатых: 7.56%
Случаи задолжностей к погашениям без задолжностей в гражданском браке: 9.31%
Случаи задолжностей к погашениям без задолжностей для вдов и вдовцев : 6.62%
Случаи задолжностей к погашениям без задолжностей для тех, кто в разводе: 7.06%
Случаи задолжностей к погашениям без задолжностей для не женатых/не замужних: 9.76%


**Вывод:** 
Более ответственны за кредит оказались люди - вдовцы,вдовы. 
А хуже всего относятся к своим долговым обязательствам люди без своей семьи(не женатые, не замужние)

#### 3.3 Есть ли зависимость между уровнем дохода и возвратом кредита в срок?

In [26]:
data.head(20)

children  days_employed  dob_years            education  education_id  \
0          1    8437.673028         42               высшее             0   
1          1    4024.803754         36              среднее             1   
2          0    5623.422610         33              среднее             1   
3          3    4124.747207         32              среднее             1   
4          0  340266.072047         53              среднее             1   
5          0     926.185831         27               высшее             0   
6          0    2879.202052         43               высшее             0   
7          0     152.779569         50              среднее             1   
8          2    6929.865299         35               высшее             0   
9          0    2188.756445         41              среднее             1   
10         2    4171.483647         36               высшее             0   
11         0     792.701887         40              среднее             1   
12         0  365286.622650         65              среднее             1   
13         0    1846.641941         54  неоконченное высшее             2   
14         0    1844.956182         56               высшее             0   
15         1     972.364419         26              среднее             1   
16         0    1719.934226         35              среднее             1   
17         0    2369.999720         33               высшее             0   
18         0  400281.136913         53              среднее             1   
19         0   10038.818549         48              среднее             1   

       family_status  family_status_id gender income_type  debt  total_income  \
0    женат / замужем                 0      F   сотрудник     0        253875   
1    женат / замужем                 0      F   сотрудник     0        112080   
2    женат / замужем                 0      M   сотрудник     0        145885   
3    женат / замужем                 0      M   сотрудник     0        267628   
4   гражданский брак                 1      F   пенсионер     0        158616   
5   гражданский брак                 1      M   компаньон     0        255763   
6    женат / замужем                 0      F   компаньон     0        240525   
7    женат / замужем                 0      M   сотрудник     0        135823   
8   гражданский брак                 1      F   сотрудник     0         95856   
9    женат / замужем                 0      M   сотрудник     0        144425   
10   женат / замужем                 0      M   компаньон     0        113943   
11   женат / замужем                 0      F   сотрудник     0         77069   
12  гражданский брак                 1      M   пенсионер     0        118514   
13   женат / замужем                 0      F   сотрудник     0        130458   
14  гражданский брак                 1      F   компаньон     1        165127   
15   женат / замужем                 0      F   сотрудник     0        116820   
16   женат / замужем                 0      F   сотрудник     0        289202   
17  гражданский брак                 1      M   сотрудник     0         90410   
18    вдовец / вдова                 2      F   пенсионер     0         56823   
19         в разводе                 3      F   сотрудник     0        242831   

                                   purpose total_income_category  \
0                            покупка жилья                     B   
1                  приобретение автомобиля                     C   
2                            покупка жилья                     C   
3               дополнительное образование                     B   
4                          сыграть свадьбу                     C   
5                            покупка жилья                     B   
6                        операции с жильем                     B   
7                              образование                     C   
8                    на проведение свадьбы                     C   
9   

In [27]:
# Ваш код будет здесь. Вы можете создавать новые ячейки.
data_total_income_category = data.groupby("total_income_category")["debt"].mean()
data.groupby("total_income_category").agg({"debt":["count","mean"]})

debt          
                       count      mean
total_income_category                 
A                         25  0.080000
B                       5014  0.070602
C                      15921  0.084982
D                        349  0.060172
E                         22  0.090909

In [28]:
print(f"Случаи задолжностей к погашениям без задолжностей для людей с категорией доходности A : {('%.2f' % (data_total_income_category[0] * 100))}%")
print(f"Случаи задолжностей к погашениям без задолжностей для людей с категорией доходности B : {('%.2f' % (data_total_income_category[1] * 100))}%")
print(f"Случаи задолжностей к погашениям без задолжностей для людей с категорией доходности C : {('%.2f' % (data_total_income_category[2] * 100))}%")
print(f"Случаи задолжностей к погашениям без задолжностей для людей с категорией доходности D : {('%.2f' % (data_total_income_category[3] * 100))}%")
print(f"Случаи задолжностей к погашениям без задолжностей для людей с категорией доходности E : {('%.2f' % (data_total_income_category[4] * 100))}%")

Случаи задолжностей к погашениям без задолжностей для людей с категорией доходности A : 8.00%
Случаи задолжностей к погашениям без задолжностей для людей с категорией доходности B : 7.06%
Случаи задолжностей к погашениям без задолжностей для людей с категорией доходности C : 8.50%
Случаи задолжностей к погашениям без задолжностей для людей с категорией доходности D : 6.02%
Случаи задолжностей к погашениям без задолжностей для людей с категорией доходности E : 9.09%


**Вывод:** 
Лучше всего выплачивают кредит люди с категорией доходности D, а хуже всего выплачивают люди с категорией доходности E.
Однако случаев людей с категорией доходности E всего 22, что очень сильно может сказаться на качественности анализа. 
Такая же ситуация для людей с категорией доходности A.


#### 3.4 Как разные цели кредита влияют на его возврат в срок?

In [29]:
# Ваш код будет здесь. Вы можете создавать новые ячейки.
data_purpose_category = data.groupby("purpose_category")["debt"].mean()
data.groupby("purpose_category").agg({"debt":["count","mean"]})

debt          
                          count      mean
purpose_category                         
операции с автомобилем     4279  0.093480
операции с недвижимостью  10751  0.072551
получение образования      3988  0.092528
проведение свадьбы         2313  0.079118

In [30]:
print(f"Случаи задолжностей к погашениям без задолжностей для людей берущих кредит на операции с автомобилем : {('%.2f' % (data_purpose_category[0] * 100))}%")
print(f"Случаи задолжностей к погашениям без задолжностей для людей берущих кредит на операции с недвижимостью : {('%.2f' % (data_purpose_category[1] * 100))}%")
print(f"Случаи задолжностей к погашениям без задолжностей для людей берущих кредит на получение образования : {('%.2f' % (data_purpose_category[2] * 100))}%")
print(f"Случаи задолжностей к погашениям без задолжностей для людей берущих кредит на проведение свадьбы : {('%.2f' % (data_purpose_category[3] * 100))}%")

Случаи задолжностей к погашениям без задолжностей для людей берущих кредит на операции с автомобилем : 9.35%
Случаи задолжностей к погашениям без задолжностей для людей берущих кредит на операции с недвижимостью : 7.26%
Случаи задолжностей к погашениям без задолжностей для людей берущих кредит на получение образования : 9.25%
Случаи задолжностей к погашениям без задолжностей для людей берущих кредит на проведение свадьбы : 7.91%


**Вывод:** 
Лучше всего отдают кредит люди, берущие кредит для операций с недвижимостью.
Хуже всего отдают кредит люди, которые берут кредит для получения образования.

#### 3.5 Приведите возможные причины появления пропусков в исходных данных.

*Ответ:* Для данной таблицы пропуски были в столбцах трудового стажа и дохода.
Скорее всего эти поля были необязательными для заполнения в анкетах кредитуемых.

#### 3.6 Объясните, почему заполнить пропуски медианным значением — лучшее решение для количественных переменных.

*Ответ:* Медиана - это более качественная характеристика, которая позволяет более точно подходить к анализу. 
На нее меньше всего влияют случайные аномалии.


### Шаг 4: общий вывод.

В исследовании была проведена предообработка данных: удалили пропуски, обработали аномалии, изменили типы данных на корректные и удобные для текущей работы, обработали дубликаты и категоризировали доход по категориям в зависимости от значений дохода, а также категоризировали различные "поводы" для взятия кредита на более общие для удобства работы.
Изучив данные, выясняли, как различные условия влияют на возврат кредита.
По категориям более ответственными при возврате кредита оказались люди: которые не имеют детей, вдовцы(вдовы), с категорией доходности D (с уровнем дохода более 50к рублей и до 200к рублей) и люди, которые берут кредит для операций с недвижимостью.
Также было выяснено, что хуже всего отдают кредит люди: которые имеют 4 детей (но данные имеют малочисленное кол-во для качественного анализа), не женатых, с категорией доходности E - доходность до 30к рублей  (также малочисленное кол-во для качественного анализа) и для операций с автомобилем.